pkg/chunk/cache/fifo_cache_test.go

```diff
@@ -2,100 +2,161 @@ package cache
-	"strconv"
+	"fmt"
 	"testing"
 	"time"
-	"unsafe"
 
 	"github.com/prometheus/client_golang/prometheus/testutil"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-const size = 10
-const overwrite = 5
+func TestFifoCacheEviction(t *testing.T) {
+	const (
+		cnt       = 10
+		overwrite = 5
+	)
+	itemTemplate := &cacheEntry{
+		key:   "00",
+		value: 0,
+	}
+	testFifoCacheEviction(t, "test-memory-eviction", FifoCacheConfig{MaxSizeBytes: cnt * sizeOf(itemTemplate), Validity: 1 * time.Minute})
+	testFifoCacheEviction(t, "test-items-eviction", FifoCacheConfig{MaxSizeItems: cnt, Validity: 1 * time.Minute})
+}
 
-func TestFifoCache(t *testing.T) {
-	c := NewFifoCache("test1", FifoCacheConfig{Size: size, Validity: 1 * time.Minute})
+func testFifoCacheEviction(t *testing.T, testname string, cfg FifoCacheConfig) {
+	const (
+		cnt       = 10
+		overwrite = 5
+	)
+	itemTemplate := &cacheEntry{
+		key:   "00",
+		value: 0,
+	}
+	c := NewFifoCache(testname, FifoCacheConfig{MaxSizeItems: cnt, Validity: 1 * time.Minute})
 	ctx := context.Background()
 
 	// Check put / get works
 	keys := []string{}
 	values := []interface{}{}
-	for i := 0; i < size; i++ {
-		keys = append(keys, strconv.Itoa(i))
+	for i := 0; i < cnt; i++ {
+		keys = append(keys, fmt.Sprintf("%02d", i))
 		values = append(values, i)
 	}
 	c.Put(ctx, keys, values)
-	require.Len(t, c.index, size)
-	require.Len(t, c.entries, size)
+	require.Len(t, c.entries, cnt)
+
+	assert.Equal(t, testutil.ToFloat64(c.entriesAdded), float64(1))
+	assert.Equal(t, testutil.ToFloat64(c.entriesAddedNew), float64(cnt))
+	assert.Equal(t, testutil.ToFloat64(c.entriesEvicted), float64(0))
+	assert.Equal(t, testutil.ToFloat64(c.entriesCurrent), float64(cnt))
+	assert.Equal(t, testutil.ToFloat64(c.totalGets), float64(0))
+	assert.Equal(t, testutil.ToFloat64(c.totalMisses), float64(0))
+	assert.Equal(t, testutil.ToFloat64(c.staleGets), float64(0))
+	assert.Equal(t, testutil.ToFloat64(c.memoryBytes), float64(cnt*sizeOf(itemTemplate)))
 
-	for i := 0; i < size; i++ {
-		value, ok := c.Get(ctx, strconv.Itoa(i))
+	for i := 0; i < cnt; i++ {
+		value, ok := c.Get(ctx, fmt.Sprintf("%02d", i))
 		require.True(t, ok)
 		require.Equal(t, i, value.(int))
 	}
 
+	assert.Equal(t, testutil.ToFloat64(c.entriesAdded), float64(1))
+	assert.Equal(t, testutil.ToFloat64(c.entriesAddedNew), float64(cnt))
+	assert.Equal(t, testutil.ToFloat64(c.entriesEvicted), float64(0))
+	assert.Equal(t, testutil.ToFloat64(c.entriesCurrent), float64(cnt))
+	assert.Equal(t, testutil.ToFloat64(c.totalGets), float64(cnt))
+	assert.Equal(t, testutil.ToFloat64(c.totalMisses), float64(0))
+	assert.Equal(t, testutil.ToFloat64(c.staleGets), float64(0))
+	assert.Equal(t, testutil.ToFloat64(c.memoryBytes), float64(cnt*sizeOf(itemTemplate)))
+
 	// Check evictions
 	keys = []string{}
 	values = []interface{}{}
-	for i := size; i < size+overwrite; i++ {
-		keys = append(keys, strconv.Itoa(i))
+	for i := cnt; i < cnt+overwrite; i++ {
```

Comment:
We should start at `cnt-overwrite` otherwise we miss to check some times. Then the check below on `c.totalGets` should adapted.

```suggestion
	for i := cnt-overwrite; i < cnt+overwrite; i++ {
```